lib/Sema/SemaCast.cpp

```diff
@@ -2945,20 +2945,80 @@ ExprResult Sema::BuildCheriCast(SourceLocation LParenLoc,
     if (Context.getTargetInfo().areAllPointersCapabilities()) {
       assert(SrcIsCap && DestIsCap);
     } else {
-      Diag(KeywordLoc, diag::warn_cheri_cast_noop) << SrcTy << DestTy
+      Diag(KeywordLoc, diag::warn_cheri_ptr_noop) << SrcTy << DestTy
         << FixItHint::CreateRemoval(SourceRange(LParenLoc, RParenLoc));
     }
   }
   return CStyleCastExpr::Create(Context, DestTy, VK_RValue, Kind, SubExpr,
                                 nullptr, TSInfo, LParenLoc, RParenLoc);
 }
 
-ExprResult Sema::ActOnCheriCast(Scope *S, SourceLocation LParenLoc,
+ExprResult Sema::BuildCheriOffsetOrAddress(SourceLocation LParenLoc,
+                                           SourceLocation KeywordLoc,
+                                           tok::TokenKind Kind, QualType
+                                           DestTy, TypeSourceInfo *TSInfo,
+                                           SourceLocation RParenLoc, Expr
+                                           *SubExpr) {
+  // Use getRealReferenceType() because getType() returns T for T&
+  QualType SrcTy = SubExpr->getRealReferenceType();
+  bool SrcIsCap = SrcTy->isCHERICapabilityType(Context);
+  if (!SrcIsCap) {
+    // XXXKG: What about functions?
+    Diag(SubExpr->getLocStart(), diag::err_cheri_offset_addr_invalid_source_type)
+      << SrcTy << (Kind == tok::kw___cheri_offset);
+    return ExprError();
+  }
+
+  // XXXKG: Should pointer types also be allowed for __cheri_addr?
```

Comment:
No, but a more-specific warning would be good (e.g. `__cheri_addr` is not valid for integral pointer types).